### PR TITLE
fix(security): harden collections & metadata editing endpoints

### LINKED ISSUE #659: Collections access control & security review
## Summary

Security review and hardening for the user document collections feature. Ensure proper access control, input validation, and rate limiting on all collections endpoints.

**PRD Reference:** `docs/prd/user-document-collections.md` — Sections 3.2, 8 (Phase 2)

### Access Control
- Verify all collection endpoints enforce user ownership (user_id from JWT must match collection owner)
- Unauthorized access to other users' collections returns 403/404
- Ensure cascading deletes respect ownership boundaries

### Input Validation
- Collection name: 1-255 characters, required, non-empty after trim
- Description: ≤1000 characters
- Note text: ≤5000 characters
- Document IDs: validated against expected format
- Position (reorder): non-negative integer
- Bulk operations: max 50 items per request

### Rate Limiting
- Read endpoints: 100/minute (standard)
- Write endpoints: 10/minute (collections), 20/minute (items)
- Bulk add operations: covered by items rate limit

### SQL Injection Prevention
- Review all SQLite queries for parameterized query usage
- Ensure no string interpolation in SQL statements

### Security Test Cases
- Unauthorized access attempts (wrong user, no auth, expired token)
- Boundary condition tests (max length strings, edge-case positions)
- Injection attempts in name, description, note fields

## Acceptance Criteria
- [ ] All endpoints enforce user ownership — users cannot view/edit other users' collections
- [ ] Input validation rejects out-of-bounds values with clear error messages
- [ ] Rate limiting configured on all write endpoints
- [ ] No SQL injection vulnerabilities (parameterized queries throughout)
- [ ] Security test cases written and passing
- [ ] No new Bandit warnings
- [ ] Review report documenting findings

## Dependencies
- Blocks frontend work (Phase 3/4)
- Depends on: Collections backend API (Phase 1)

## Notes
- Can be merged into Phase 1 PR or as separate PR (`squad/591-phase2-access-control`)
- Review authority: Kane (Security Engineer)

### LINKED ISSUE #695: Metadata editing security review (admin auth, input sanitization, Redis access)
## Summary

Security review for the book metadata editing feature: admin auth enforcement, input sanitization, and Redis access control.

**PRD Reference:** `docs/prd/book-metadata-editing.md` — Sections 3.4, 7 (Phase 4)

### Admin Auth Validation
- Verify all metadata edit endpoints require admin role (`admin_auth.py`)
- Non-admin users get 403 Forbidden
- Token expiration handled correctly
- Test with various auth scenarios (no token, expired token, non-admin token, admin token)

### Input Sanitization
- Solr query injection prevention in batch-by-query endpoint (`query` parameter)
- String field sanitization (title, author, category, series)
- Year range validation (1000-2099)
- Reject empty-after-trim strings
- Max length enforcement on all fields

### Redis Access Control
- Review Redis key pattern: `aithena:metadata-override:{doc_id}`
- Ensure no key enumeration is possible from API
- Verify permanent TTL behavior (no expiration)
- Check for Redis command injection in key construction

### Solr Update Security
- Verify atomic update requests are properly constructed (no field injection)
- Ensure only allowed fields can be updated via API

## Acceptance Criteria
- [ ] All metadata edit endpoints enforce admin role
- [ ] Non-admin access attempts return 403 with no data leakage
- [ ] Input validation prevents injection in all fields
- [ ] Solr query parameter in batch-by-query is safely handled
- [ ] Redis key construction is safe from injection
- [ ] No new Bandit warnings
- [ ] Security review report documenting findings and recommendations

## Dependencies
- Depends on: Single edit API (#681) and Batch edit API (#683) — code must exist to review

### PR DESCRIPTION
**Pull request by jmservera**
## Security Review & Hardening — Collections (#659) + Metadata Editing (#695)

Working as **Kane** (Security Engineer)

### Vulnerabilities Fixed

1. **Metadata edit endpoint missing JWT role check** (Critical)
   - The PATCH `/v1/admin/documents/{doc_id}/metadata` endpoint only required an API key but did NOT verify the user had admin role via JWT
   - Any authenticated user with the API key could edit metadata regardless of role
   - **Fix:** Added `require_role("admin")` as additional dependency (defense-in-depth)

2. **Collection item position allowed negative values**
   - `UpdateItemRequest.position` had no constraint — negative values could cause unexpected ordering
   - **Fix:** Added `Field(ge=0)` Pydantic constraint

3. **Reorder request had no list size limit**
   - `ReorderItemsRequest.item_ids` had no `max_length` — potential DoS with huge lists
   - **Fix:** Added `max_length=50` (consistent with `AddItemsRequest`)

4. **Collection item note lacked Pydantic-level validation**
   - Note was only validated at endpoint level, not at model level
   - **Fix:** Added `Field(max_length=5000)` for defense-in-depth

### Security Tests Added (67 new tests)

| Category | Count |
|----------|-------|
| Cross-user access control isolation | 8 |
| Unauthenticated/expired token rejection | 3 |
| Collection not-found handling | 3 |
| Input validation boundaries | 10 |
| Bulk operation size limits | 4 |
| SQL injection prevention | 5 |
| Cascading delete ownership | 2 |
| Admin auth (role + API key + token scenarios) | 8 |
| Solr query injection prevention | 6 |
| Metadata field validation boundaries | 12 |
| Redis key construction safety | 3 |
| Solr atomic update security | 2 |
| XSS payload behavior | 1 |

### Test Results
- All 664 tests pass (597 existing + 67 new)
- Coverage: 94.72% (threshold: 88%)
- Bandit: Clean (only pre-existing B104 finding)

### Files Changed
- `src/solr-search/main.py` — Added `require_role("admin")` to metadata edit endpoint
- `src/solr-search/collections_models.py` — Hardened Pydantic model validation
- `src/solr-search/tests/test_collections_security.py` — 35 security tests (NEW)
- `src/solr-search/tests/test_metadata_edit_security.py` — 32 security tests (NEW)

Closes #659
Closes #695

### CODE REVIEW

File: src/solr-search/tests/test_metadata_edit_security.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        with (
+            patch("main._get_redis_pool"),
+            patch("main._raw_solr_query", return_value={"response": {"numFound": 0, "docs": []}}) as mock_query,
+        ):
+            client = _client_for(ADMIN_USER)
+            resp = client.patch("/v1/admin/documents/*/metadata", json={"title": "Wildcard"})
+        assert resp.status_code == 404  # noqa: S101
+        # Verify the query passed to Solr has the wildcard escaped
+        call_args = mock_query.call_args[0][0]
+        assert "\\*" in call_args.get("q", "") or resp.status_code == 404  # noqa: S101
```
[the remark this answers is not in the document]
Comment (jmservera) [reply src/solr-search/tests/test_metadata_edit_security.py]: Good catch — the `or resp.status_code == 404` was always true (tautological). Reworked to verify mock_query was actually called and the wildcard was either escaped or not passed raw.

File: src/solr-search/tests/test_metadata_edit_security.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+        client = _client_for(ADMIN_USER)
+        resp = client.patch(ENDPOINT, json={})
+        assert resp.status_code == 422  # noqa: S101
+
+    def test_unknown_fields_ignored(self):
+        """Extra fields not in _METADATA_FIELD_MAP should not cause errors."""
+        client = _client_for(ADMIN_USER)
+        resp = client.patch(ENDPOINT, json={"unknown_field": "value"})
+        assert resp.status_code == 422  # noqa: S101 (no valid fields)
+
```
[the remark this answers is not in the document]
Comment (jmservera) [reply src/solr-search/tests/test_metadata_edit_security.py]: Fixed — renamed to `test_unknown_fields_rejected` with updated docstring to match the 422 behavior.
